Controllers/CHumidityController.cpp

```diff
@@ -0,0 +1,195 @@
+/*
+ * CHumidityController.cpp
+ *
+ *  Created on: Aug. 11, 2022
+ *      Author: samk
+ */
+
+#include "CHumidityController.h"
+#include <stdio.h>
+#include "CBME280.h"
+#include "CHumidifier.h"
+#include "ICommand.h"
+#include "IHardwareMap.h"
+
+#define MIN_TEMPERATURE  10
+#define MAX_TEMPERATURE  50
+#define MIN_POWER        0
+#define MAX_POWER        100
+#define DISABLE_OVERRIDE -1
+#define DISABLE_TARGET   0
+#define MIN_HUMIDITY     85
+#define MAX_HUMIDITY     95
+
+CHumidityController::CHumidityController(IHardwareMap *p_hardware,
+                                         etl::string<MAX_STRING_SIZE> name,
+                                         uint32_t run_period_ms,
+                                         SPI_HandleTypeDef *p_spi,
+                                         GPIO_TypeDef *p_slave_select_port,
+                                         uint16_t slave_select_pin)
+    : CController(name, run_period_ms),
+      mp_hw(p_hardware)
+{
+    // TODO Auto-generated constructor stub
+    reset();
+
+    CBME280 humidity_sensor(p_spi, p_slave_select_port, slave_select_pin);
+    m_humidity_sensor = &humidity_sensor;
+}
+
+/**
+ * @brief Run all repeated humidity control activities.
+ *
+ */
+void CHumidityController::run()
+{
+    float power = 0;
+    if (m_power_override == DISABLE_OVERRIDE)
+    {
+        if (m_target_humidity != DISABLE_TARGET)
+        {
+            float actual_humidity = mp_humidity_sensor->getHumidity();
+            if ((MIN_HUMIDITY <= actual_humidity) &&
+                (actual_humidity <= MAX_humidity))
+            {
+                power = m_control_loop.run(m_target_humidity, actual_humidity);
+            }
+        }
+    }
+    else
+    {
+        power = m_power_override;
+    }
+    mp_hw->setHumidifierPower(power);
+}
+
+bool CHumidityController::newCommand(ICommand *p_command,
+                                     IComChannel *p_comchannel)
+{
+    bool b_command_recognised = false;
+    ICommand::command_error_code_t result = ICommand::COMMAND_OK;
+    if (p_command->getName()->compare("?humidity") == 0)
+    {
+        sendStatus(p_comchannel);
+        b_command_recognised = true;
+    }
+
+    if (p_command->getName()->compare("humidity") == 0)
+    {
+        result = setHumidity(p_command);
+        b_command_recognised = true;
+    }
+
+    if (p_command->getName()->compare("heater") == 0)
```

Comment:
Should probably have a more obvious name for a command that overrides humidity power. Perhaps: humidifier_override().